### PR TITLE
Enhancement: Support hashing hostname and filename

### PR DESCRIPTION
**Pull request by hariharan-devarajan**
1. We hash the filename and create a metadata event on load.
2. On opening a file, we store the hash instead of the whole value.
3. Use a md5 hash implementation for consistent hashing.
4. Added `dftracer_anonymize` script to ignore all hash events and anonymize the trace filenames.
5. Support for the end symbol for perfetto was added.
6. Support to have exec and cmd as hash as well.
7. Using H for phase event for hash events.
8. Using M for phase events for metadata events.
9. Added support for core-affinity analysis.

### CODE REVIEW

File: src/dftracer/writer/chrome_writer.cpp [hunk 1]
```diff
@@ -200,16 +216,15 @@ void dftracer::ChromeWriter::convert_json(
         case EventType::COMPLETE_EVENT: {
           auto written_size = sprintf(
               buffer.data() + current_index,
-              R"(%s{"id":%d,"name":"%s","cat":"%s","pid":%lu,"tid":%lu,"ts":%llu,"dur":%llu,"ph":"X","args":{"hostname":"%s"%s}})",
+              R"(%s{"id":%d,"name":"%s","cat":"%s","pid":%lu,"tid":%lu,"ts":%llu,"dur":%llu,"ph":"X","args":{"hhash":%d%s}})",
```
Comment (rayandrew) [inline src/dftracer/writer/chrome_writer.cpp:219]: is this mistyped?

**Review by rayandrew — CHANGES_REQUESTED** [review]

**Comment by rayandrew** [comment]
it seems the `cmd` is not hashed. should we make it hashable or no?

```
{"id":1,"name":"start","cat":"dftracer","pid":715314,"tid":1430628,"ts":1727232336480666,"dur":0,"ph":"X","args":{"hhash":48639,"p_idx":-1,"level":1,"ppid":715312,"date":"Wed Sep 25 02:45:36 2024","cmd":"/lus/eagle/projects/MDClimSim/rayandrew/stormer-dev-pure/venvs/2024-04-29/bin/python3;-Wignore;/lus/eagle/projects/MDClimSim/rayandrew/stormer-dev-pure/stormer.py;--strategy;fsdp;--num-heads;16;--hidden-depth;24;--hidden-size;1024;--patch-size;10;--mlp-ratio;4.0;--weight-decay;0;--lr;1e-4;--prefetch-factor;256;--num-epochs;5;--persistent-workers;--val-lead-times;6,72;--list-train-intervals;6;--num-workers;6;--val-batch-size;1;--batch-size;1;--data-frequency;6;--year-list;1979,1980;--checkpoint-dir;/lus/eagle/projects/MDClimSim/rayandrew/experiments/stormer/checkpoints/polaris_jobid-2097079_nh-1_ws-4_dtype-bf16_ts-1727232275_date-20240925;--data-dir;/eagle/MDClimSim/rayandrew/data/striped;--output-dir;/lus/eagle/projects/MDClimSim/rayandrew/experiments/stormer/logs/polaris_jobid-2097079_nh-1_ws-4_dtype-bf16_ts-1727232275_date-20240925","exec":"stormer.py","version":100005}}
...

```

**Comment by rayandrew** [comment]
Also, I cannot see any POSIX events on perfetto anymore


image

**Comment by hariharan-devarajan** [comment]
Can u search for open I was able to see them

**Comment by hariharan-devarajan** [comment]
> it seems the `cmd` is not hashed. should we make it hashable or no?
> 
> 
> 
> ```
> 
> {"id":1,"name":"start","cat":"dftracer","pid":715314,"tid":1430628,"ts":1727232336480666,"dur":0,"ph":"X","args":{"hhash":48639,"p_idx":-1,"level":1,"ppid":715312,"date":"Wed Sep 25 02:45:36 2024","cmd":"/lus/eagle/projects/MDClimSim/rayandrew/stormer-dev-pure/venvs/2024-04-29/bin/python3;-Wignore;/lus/eagle/projects/MDClimSim/rayandrew/stormer-dev-pure/stormer.py;--strategy;fsdp;--num-heads;16;--hidden-depth;24;--hidden-size;1024;--patch-size;10;--mlp-ratio;4.0;--weight-decay;0;--lr;1e-4;--prefetch-factor;256;--num-epochs;5;--persistent-workers;--val-lead-times;6,72;--list-train-intervals;6;--num-workers;6;--val-batch-size;1;--batch-size;1;--data-frequency;6;--year-list;1979,1980;--checkpoint-dir;/lus/eagle/projects/MDClimSim/rayandrew/experiments/stormer/checkpoints/polaris_jobid-2097079_nh-1_ws-4_dtype-bf16_ts-1727232275_date-20240925;--data-dir;/eagle/MDClimSim/rayandrew/data/striped;--output-dir;/lus/eagle/projects/MDClimSim/rayandrew/experiments/stormer/logs/polaris_jobid-2097079_nh-1_ws-4_dtype-bf16_ts-1727232275_date-20240925","exec":"stormer.py","version":100005}}
> 
> ...
> 
> 
> 
> ```

That's a good idea. I will work on that

**Comment by rayandrew** [comment]
> Can u search for open I was able to see them

I cannot see it in perfetto but yes I can search it using grep.

That is weird

**Comment by rayandrew** [comment]
I added fix for merge script where there are additional `][` and `]` at the beginning of the line

**Comment by rayandrew** [comment]
I think we should make the hash for `process_name` and `exec` at the beginning of the dftracer execution since this sometimes contain sensitive information

Here is the example

```
{"id":2,"name":"mpi","cat":"dftracer","pid":211129,"tid":422258,"ts":1727163745895128,"dur":0,"ph":"X","args":{"hostname":"x3209c0s37b0n0","rank":0}}
{"id":3,"name":"process_name","pid":211129,"tid":422258,"ph":"M","args":{"name":"Rank 0"}}
{"id":1,"name":"start","cat":"dftracer","pid":211129,"tid":422258,"ts":1727163745895126,"dur":0,"ph":"X","args":{"hostname":"x3209c0s37b0n0","p_idx":-1,"level":1,"ppid":211127,"date":"Tue Sep 24 07:42:25 2024","cmd":"/lus/eagle/projects/MDClimSim/rayandrew/stormer-dev-pure/venvs/2024-04-29/bin/python3;-Wignore;/lus/eagle/projects/MDClimSim/rayandrew/stormer-dev-pure/stormer.py;--strategy;none;--num-heads;16;--hidden-depth;24;--hidden-size;1024;--patch-size;16;--mlp-ratio;4.0;--weight-decay;0;--lr;1e-4;--prefetch-factor;128;--num-epochs;5;--persistent-workers;--val-lead-times;6,72;--list-train-intervals;6;--num-workers;10;--val-batch-size;1;--batch-size;1;--data-frequency;6;--year-list;1979;--checkpoint-dir;/lus/eagle/projects/MDClimSim/rayandrew/experiments/stormer/checkpoints/polaris_jobid-2095939_nh-1_ws-4_dtype-bf16_ts-1727163735_date-20240924;--data-dir;/eagle/MDClimSim/rayandrew/data/striped;--output-dir;/lus/eagle/projects/MDClimSim/rayandrew/experiments/stormer/logs/polaris_jobid-2095939_nh-1_ws-4_dtype-bf16_ts-1727163735_date-20240924","exec":"stormer.py","version":100004}}
```

**Comment by hariharan-devarajan** [comment]
This PR adds hash values instead of strings within the code. Example of these cases are
Start event has cmd and exec hash
```
{"id":1,"name":"/usr/workspace/haridev/dftracer/build/bin/test_cpp;/usr/workspace/haridev/dftracer/build/test/data","cat":"dftracer","pid":3147907,"tid":3147907,"ph":"H","args":{"hash":22138}}
{"id":2,"name":"test_cpp","cat":"dftracer","pid":3147907,"tid":3147907,"ph":"H","args":{"hash":355}}
{"id":3,"name":"start","cat":"dftracer","pid":3147907,"tid":6295814,"ts":1727242900616194,"dur":0,"ph":"X","args":{"hhash":39537,"p_idx":-1,"cmd_hash":22138,"exec_hash":355,"level":1,"version":100005,"date":"Tue Sep 24 22:41:40 2024","ppid":3147902}}
```

Similarly, hostnames are now hashed.
```
{"id":0,"name":"corona82","cat":"dftracer","pid":3147907,"tid":3147907,"ph":"H","args":{"hash":39537}}
{"id":9,"name":"truncate","cat":"POSIX","pid":3147907,"tid":6295814,"ts":1727242903627805,"dur":535,"ph":"X","args":{"hhash":39537,"p_idx":-1,"level":1,"fhash":27060,"length":0}}
```

Finally, file paths are hashed.
```
{"id":8,"name":"/usr/workspace/haridev/dftracer/build/test/data/demofile.txt","cat":"dftracer","pid":3147907,"tid":3147907,"ph":"H","args":{"hash":27060}}
{"id":9,"name":"truncate","cat":"POSIX","pid":3147907,"tid":6295814,"ts":1727242903627805,"dur":535,"ph":"X","args":{"hhash":39537,"p_idx":-1,"level":1,"fhash":27060,"length":0}}
{"id":10,"name":"fopen","cat":"STDIO","pid":3147907,"tid":6295814,"ts":1727242903628501,"dur":341,"ph":"X","args":{"hhash":39537,"level":1,"fhash":27060,"p_idx":-1,"mode":"w+"}}
```

**Comment by hariharan-devarajan** [comment]
These changes will have the following effect
1. Traces are more compact.
2. It is easy to anonymize the scripts by simply remove the `ph=H` events.

**Comment by rayandrew** [comment]
There is one quirk that I still do not know how to fix

If I anonymize the trace and merge it then visualize in perfetto, every event seems doubled

image

**Comment by hariharan-devarajan** [comment]
Does the trace have double events?

**Comment by hariharan-devarajan** [comment]
And does it happen when do not anonymize?

**Review by rayandrew — APPROVED** [review]
looks good just tested it in my project!

**Pull request merged**